flexget/components/estimate_release/estimate_release.py

```diff
@@ -12,32 +34,105 @@ class EstimateRelease:
     for various things (series, movies).
     """
 
-    def estimate(self, entry):
+    def __init__(self):
+        self.task_estimate_config = {}
+
+    @property
+    def schema(self) -> Dict[str, Any]:
+        """Create schema that allows configuring estimator providers and
+        related settings.
+        """
+
+        schema = {
+            'type': 'object',
+            'properties': {
+                'providers': {
+                    'type': 'array',
+                    'items': {
+                        'type': 'string',
+                        'enum': [
+                            p.name.replace('est_', '')
+                            for p in plugin.get_plugins(interface=ESTIMATOR_INTERFACE)
+                        ],
+                    },
+                },
+                'offset': {'type': 'string', 'format': 'interval', 'default': '0 days'},
+            },
+            'additionalProperties': False,
+        }
+
+        return schema
+
+    def on_task_start(self, task: Task, config) -> None:
+        # Load task specific estimator configuration
+        if config:
+            self.task_estimate_config = config
+
+    def on_task_exit(self, task: Task, config) -> None:
+        # Restore default estimator configuration for next task run
+        self.task_estimate_config = {}
```

Comment:
This modifying configuration and restoring later is not a good idea. Generally speaking plugins should be stateless and not rely on __init__.

Reply:
Interesting, this is how the `parsing` plugin does it which was mentioned in one of the above comments as a reference.

Reply:
Hmm, apologies. I guess there's no better way for it in these kinds of situations then. There is still issue with having `self.task_estimate_config` and the one declared in module level (L18). I don't think the module one is used and can be removed.

Reply:
Yeah, we have to do some hacky stuff for bits that don't directly get a task parameter, there isn't a better way at the moment. I was thinking of making a plugin method that worked like a context manager, to make plugins that do setup and teardown around tasks a bit easier. Something like:
```python
def on_task_context(self, task, config):
    if config:
        self.task_estimate_config = config
    try:
        yield
    finally:
        self.task_estimate_config = {}
```
That way you don't have to hook task_start, task_exit, and task_abort trying to manage state. @paranoidi Thoughts?

I also had this crazier idea based on that, but my gut tells me it's probably bad. It would allow a plugin to retain state during a task without storing things to self.
```python
def on_task_phases(self, task, config):
    yield 'start'
    setupstuff = somefunc()
    yield 'filter'
    for entry in task.all_entries:
        do_something(entry, setupstuff)
    yield 'exit'
    teardown(setupstuff)
```

Reply:
Heh, sure. I have some thoughts. Not overly sold on the idea. 

I've been thinking years that we should have a baseclass for plugins and in there have a `prepare_config` method that plugins can override to format config with default values etc. Then passing the already prepared config into the usual phase methods. Also this would make it possible to declare the phase methods in the baseclass making them much more explicit. Last time I checked there was only one dynamically added phase which would not be a hard to get rid of.

Reply:
> I've been thinking years that we should have a baseclass for plugins and in there have a `prepare_config` method that plugins can override to format config with default values etc. Then passing the already prepared config into the usual phase methods.

Yeah, I love this idea.
I think it's totally separate than a context manager type thing to do cleanup, and I'm not totally sold on that idea either. The problem of doing setup and teardown in on_task_start/exit is that you can't guarantee that they run in pairs. I just fixed an issue with the parsing plugin where the on_task_abort cleanup ran even though the task was aborted before its on_task_start even set up the thing it was trying to clean up. You could also run into the issue of an abort happening after on_task_exit handler was run, and it tries to clean up twice. Here is a [naive search](https://github.com/search?q=repo%3AFlexget%2FFlexget+%22on_task_abort+%3D+on_task_exit%22&type=code) that shows some plugins that might be able to use a thing like the context manager hook.

Reply:
Maybe the baseclass could introduce setup/teardown methods (although I kinda dislike those names as they are heavily unit test related) that are guaranteed to run as expected. It would be much more explicit than relying on on_task_start etc.